### DISCUSSION TITLE
Forest 🌲 Roadmap Discussion [Public]

### DISCUSSION THREAD

**Discussion by zettai-caiman** [ChainSafe/forest#2172] [Forest 🌲 Q&A]
Discuss anything related to Forest 🌲 roadmap

**Comment by q9f** [comment]
> Forest :evergreen_tree: Beta will include: FIL wallet handling, storage or retrieval market

> Features: implementation of storage & retrieval markets, node wallet

FIL wallets/account management should be a separate, stand-alone binary and not by default part of the full-node client. ref https://github.com/ChainSafe/forest/issues/1517

**Reply by zettai-caiman** [reply]
Why separate the FIL wallet handing? This can be one of the main features. 

If you can't even store, manage, send, receive FIL there won't be much of a value proposition for the Forest client aka no incentive to run it at all just to maintain consensus with the network. 

Unless, we are trying to keep the Forest core super light weight? Then have FIL wallet handing or any other feature as an add on.

**Reply by q9f** [reply]
Having account management in nodes is the (arguably) number one pain point. Here's some context: https://github.com/ethereum/go-ethereum/issues/25948

Geth created an external signer (Clef) and we should have that straight from the beginning: https://geth.ethereum.org/docs/interface/managing-your-accounts

I'm not saying we shouldn't have a wallet. I'm just trying to make sure it's not part of the forest node. It should be a separate binary:
* forestd - the client, with APIs
* forest-cli - a command-line interface
* forest-account - an external signer/account management tool

Binary names are just exemplary.

**Reply by zettai-caiman** [reply]
Ah, yes ok. I understand what you mean now. @lemmih thoughts on the above?

**Comment by q9f** [comment]
> Blockchain client fundamentals: RPC endpoints, CLI command interface, Consensus layer, Networking layer

Would love to see someone building a FIL explorer MVP using Forest.

**Reply by zettai-caiman** [reply]
- Built by ChainSafe (or the Forest 🌲 team) or external third parties with Forest 🌲 API? 
- As of now, Forest 🌲 API mayn't have anything specific of value for an additional FIL block explorer. There will be a lack of differentiation.

**Reply by q9f** [reply]
It's not about differentiation, it's more about proving we are offering feature-complete services. A simple html-explorer fetching data from a forest node would be a lovely showcase.

**Reply by zettai-caiman** [reply]
Ya ok, I see. More like a demonstration of a use case from Forest API.

**Comment by q9f** [comment]
> Reliability: regular quarterly releases, automated testing and network status reporting, accurate and high-level change logs

I would propose 6-weekly releases along with the Rust compiler, it would not require a dedicated process yet but having a tag every other month would be good.

**Reply by zettai-caiman** [reply]
Currently, we doing quarterly releases. Shorter releases like 6 weeks or monthly would be ideal. However, I'm comfortable with either -at the minimum we have to do quarterly. 

Does the Rust complier do a 6 week release?

**Reply by q9f** [reply]
Yes.

**Reply by LesnyRumcajs** [reply]
I want to add my few cents here.

In general, our releases don't add any additional overhead, i.e., there's no additional QA phase to see if what wasn't covered by tests still works (say, sending FIL, managing a wallet etc.). Perhaps at this point, it is not needed, but something to keep in mind if we want to consider Forest anywhere close to prod.

If we want meaningful releases, we should have a test plan with manual checks for every feature that is not part of the CI. Hopefully, in time the number of such checks will decrease. Otherwise, such 6-week releases will not be worth anything more than a checkmark in a document, and a tagged commit will be essentially on par (in terms of how confident we are with it) with a random commit in `main` (which would be good, but only if we would have 100% coverage which is not realistic right now).

**Reply by zettai-caiman** [reply]
@LesnyRumcajs What is the current automated coverage? Perhaps we can establish a plan to include a QA phase before every release?

**Reply by LesnyRumcajs** [reply]
@lerajk It's 34% for the unit tests, and it would go higher if we included the sync checks (they aren't 100% exhaustive, though; they check a few tipsets). What we are missing the most is coverage for RPC commands, combinations of various flags and node stability. Currently, we would know that a PR increases memory usage only if Forest starts crashing.

So yeah, we could contrive a test plan to go through before each release. Again, no blockers, we only need to prioritize it and make it happen.

There's also a positive aspect of having to do manual tests if they're not too cumbersome - every Forest dev would know by heart Forest commands and flags. Provided there's not a single person doing those checks every time, but we do it on a rotation basis.

**Reply by zettai-caiman** [reply]
@q9f is there a possibly to get help from ChainSafe QA team for the above for Forest? 

@LesnyRumcajs Let us create an issue for a test plan and coverage? Perhaps it is a good idea to stay away from manual tests, an automated plan is more manageable in the long run and we don't run into issues of knowledge gaps.

**Reply by q9f** [reply]
Coverage has always been a problem and I don't think the ChainSafe QA team will be much of a help here. I would rather encourage you to do it similar to the Lodestar team and hire a "Software Engineer in Testing." Let them start writing unit tests for literally everything and slowly work their way up to integration tests and eventually more complex end-to-end simulations.

Also, manual testing is not necessarily a bad thing. Sometimes you have the most involved test-suite and at the end of the release realize you cannot "double click" your executable. I'd be happy to assist for a few release cycles.

**Reply by zettai-caiman** [reply]
Yes ok, will discuss and think about hiring a Software Engineer in Testing with the team. Test coverage will be a crucial criteria for a mainnet production node.

**Comment by LesnyRumcajs** [comment]
This is related to releases a bit.

What we are missing is a stable node. At the moment, we don't have one - only some sync checks, which are working quite well, but they are not simulating a production node. If we want Forest to have a meaningful place in the network (also related to [Network Statistics Milestone](https://github.com/ChainSafe/forest/milestone/3)), we should provide at least one "real" node, constantly up.

This would bring some benefits:
* Establish a release process, force us to document upgrade steps, develop mature practices across the team, and introduce public metrics available to all,
* Catch issues that we are potentially missing because (e.g., long-running process leaking memory, database bloat, server running out of space),
* Truly participate in the Filecoin network and be able to say that **at least** one Forest node is constantly validating blocks.

The node setup and maintenance should be as easy as in Saturn L1 https://github.com/filecoin-saturn/L1-node#set-up-a-node.

The cost of this is a VPS on Digital Ocean, roughly $100 per month unless it turns out we need a beefier machine (which would be a remarkable finding; currently, we can't recommend anything objectively). 

Ultimately, this single node should bootstrap Forest into a healthy production stage.

**Reply by zettai-caiman** [reply]
Yes, let's do this - run a permanent node in Digital Ocean. What blockers do we have now to execute the ⬆️

**Reply by LesnyRumcajs** [reply]
No blockers, only scoping and prioritizing it. And then putting someone to work on it.

**Reply by zettai-caiman** [reply]
@q9f can we get help from ChainSafe devops to run a Forest 🌲 node on an ongoing basis for the above? 

@LesnyRumcajs gonna create an issue for this as well.

**Reply by q9f** [reply]
We are currently hiring for this, yes. (Blockchain Operations).
